### ISSUE TITLE
save a game snapshot during crash reporting

### ISSUE THREAD

**Issue by melvinzhang** [magarena/magarena#627]
I just realized it will be very useful to have a game snapshot at the point of the crash, the crash reporter gets a copy of the MagicGame, so it should be possible to save it as a snapshot.

The simple hand and permanents output of the reporter is there to help us replicate the state but having a proper snapshot will be much more useful and less work required on our part as we can load the snapshot directly instead of manually reading the output of the reporter and replicating it by hand.

**Comment by ShawnieBoy** [comment]
Was getting myself confused between game snapshot and game screenshot - Did think it was already happening... :wink:

**Comment by lodici** [comment]
It might be useful to create a similar zip as is used for the gameplay report so everything is nicely packaged in one place. Both GitHub and the forum will let you attach zips directly which makes it even more convenient.

**Comment by ShawnieBoy** [comment]
Wouldn't a save-state of the crash be 'too soon'? - i.e. would probably crash when loaded. May need an undo state before save, but would prove tricky to save or do anything mid-crash.

**Comment by melvinzhang** [comment]
> would probably crash when loaded.
That is the idea! A same game we can load and trigger the crash immediately, without having to read the output of the hand/permanents and recreate it by hand.

**Comment by ShawnieBoy** [comment]
I was just thinking that we wouldn't be able to observe the crash - we'd just get the same crash log that the user sent.

**Comment by melvinzhang** [comment]
I think the main problem is right now is that sometimes we can't replicate the problem at all, being able to reliable get a crash would be very useful on its own.

**Comment by lodici** [comment]
Would the game state saved by the crash reported be the game state that caused the crash?

**Comment by ShawnieBoy** [comment]
>I think the main problem is right now is that sometimes we can't replicate the problem at all, being able to reliable get a crash would be very useful on its own.

Very true

**Comment by melvinzhang** [comment]
> Would the game state saved by the crash reported be the game state that caused the crash?

Rather it is the state at the point of the crash, which could be some half-way state not normally reachable. For example, moving a card is actually two actions, removing it from the old location, adding to the new location. If adding it to the new location generates a crash, the game state will have the card missing because it was removed but not yet added.

**Comment by melvinzhang** [comment]
> May need an undo state before save, but would prove tricky to save or do anything mid-crash.

Agree, that would be tricky. I think it is not that important for the first go at this.

The main observation is that the game report outputs the hand and battlefield permanents as part of the report txt. It is always quite painful to go through this output and manually recreate the state. The snapshot is an output in a way that can be loaded back into the game automatically and would make it a lot easier to work it. A big bonus is that the snapshot is a more complete picture, including the cards in the library and graveyard.
